esp8266/modesp.c

```diff
@@ -35,6 +35,492 @@
 #include MICROPY_HAL_H
 #include "queue.h"
 #include "user_interface.h"
+#include "espconn.h"
+
+#include "utils.h"
+
+#define IPV4ADDR_BUFSIZE    4
+
+// Helpers from cc3200/mods/modnetwork.c
+STATIC void esp_espconn_parse_ipv4_addr(mp_obj_t addr_in, uint8_t *out_ip) {
+    mp_uint_t addr_len;
+    const char *addr_str = mp_obj_str_get_data(addr_in, &addr_len);
+    if (addr_len == 0) {
+        // special case of no address given
+        memset(out_ip, 0, IPV4ADDR_BUFSIZE);
+        return;
+    }
+    const char *s = addr_str;
+    const char *s_top = addr_str + addr_len;
+    for (mp_uint_t i = 3 ; ; i--) {
+        mp_uint_t val = 0;
+        for (; s < s_top && *s != '.'; s++) {
+            val = val * 10 + *s - '0';
+        }
+        out_ip[IPV4ADDR_BUFSIZE - 1 - i] = val;
+        if (i == 0 && s == s_top) {
+            return;
+        } else if (i > 0 && s < s_top && *s == '.') {
+            s++;
+        } else {
+            nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "invalid arguments"));
+        }
+    }
+}
+
+// Takes an address of the form ('192.168.0.1', 8080), returns the port and
+// puts IP in out_ip (which must take at least IPADDR_BUF_SIZE bytes).
+STATIC mp_uint_t esp_espconn_parse_inet_addr(mp_obj_t addr_in, uint8_t *out_ip) {
+    mp_obj_t *addr_items;
+    mp_obj_get_array_fixed_n(addr_in, 2, &addr_items);
+    esp_espconn_parse_ipv4_addr(addr_items[0], out_ip);
+    return mp_obj_get_int(addr_items[1]);
+}
+
+// Takes an array with a raw IPv4 address and returns something like '192.168.0.1'.
+STATIC mp_obj_t esp_espconn_format_ipv4_addr(uint8_t *ip) {
+    char ip_str[16];
+    mp_uint_t ip_len = snprintf(ip_str, 16, "%u.%u.%u.%u", ip[0], ip[1], ip[2], ip[3]);
+    return mp_obj_new_str(ip_str, ip_len, false);
+}
+
+// Takes an array with a raw IP address, and a port, and returns a net-address
+// tuple such as ('192.168.0.1', 8080).
+STATIC mp_obj_t esp_espconn_format_inet_addr(uint8_t *ip, mp_uint_t port) {
+    mp_obj_t tuple[2] = {
+        tuple[0] = esp_espconn_format_ipv4_addr(ip),
+        tuple[1] = mp_obj_new_int(port),
+    };
+    return mp_obj_new_tuple(2, tuple);
+}
+
+STATIC const mp_obj_type_t esp_espconn_type;
+
+typedef struct _esp_espconn_obj_t {
+    mp_obj_base_t base;
+    struct espconn *espconn;
+
+    mp_obj_t cb_connect;
+    mp_obj_t cb_recv;
+    mp_obj_t cb_disconnect;
+
+    uint8_t *recvbuf;
+    mp_uint_t recvbuf_len;
+
+    bool server;
+
+    mp_obj_list_t *connlist;
+} esp_espconn_obj_t;
+
+// Due to the onconnect callback not being able to recognize the parent esp_espconn,
+// we can have only one esp_espconn listening at a time
+// This should be solvable by some PIC hacking
```

Comment:
Well, the solution could be:
1) Build the function as [position independent](http://en.wikipedia.org/wiki/Position-independent_code)
2) For every listening espconn, copy the function + some metadata to a new place in memory
3) Then, as the function gets called, find the metadata (we can calculate offset from start of function using [this](http://stackoverflow.com/questions/2154852/get-a-pointer-to-the-current-function-in-c-gcc) and get current pc using asm()) and dispatch to the appropriate callback using them
Thinking about it now, it would be way simpler to just use bare asm function though...